cpp/src/arrow/stl.h

```diff
@@ -87,33 +103,88 @@ struct ConversionTraits<std::string> : public CTypeTraits<std::string> {
   constexpr static bool nullable = false;
 };
 
-template <typename value_c_type>
-struct ConversionTraits<std::vector<value_c_type>>
-    : public CTypeTraits<std::vector<value_c_type>> {
-  static Status AppendRow(ListBuilder& builder, std::vector<value_c_type> cell) {
-    using ElementBuilderType = typename TypeTraits<
-        typename ConversionTraits<value_c_type>::ArrowType>::BuilderType;
+namespace internal {
+using ::arrow::internal::void_t;
+
+template <typename CType, typename BuilderType, typename Range, typename Enable = void>
+struct HasAppendMultipleRows : public std::false_type {};
+
+template <typename CType, typename BuilderType, typename Range>
+struct HasAppendMultipleRows<CType, BuilderType, Range,
+                             void_t<decltype(ConversionTraits<CType>::AppendMultipleRows(
+                                 std::declval<BuilderType&>(), std::declval<Range>()))>>
+    : public std::true_type {};
+
+template <typename ValueCType>
+using ElementBuilderType =
+    typename TypeTraits<typename ConversionTraits<ValueCType>::ArrowType>::BuilderType;
+
+template <typename ValueCType, typename BuilderType, typename Range,
+          typename Enable = void>
+struct AppendMultipleRowsHelper {
+  static Status impl(BuilderType& builder, Range&& cell_range) {
     ARROW_RETURN_NOT_OK(builder.Append());
-    ElementBuilderType& value_builder =
-        ::arrow::internal::checked_cast<ElementBuilderType&>(*builder.value_builder());
-    for (auto const& value : cell) {
-      ARROW_RETURN_NOT_OK(
-          ConversionTraits<value_c_type>::AppendRow(value_builder, value));
+    auto& value_builder =
+        ::arrow::internal::checked_cast<ElementBuilderType<ValueCType>&>(
+            *builder.value_builder());
+    for (auto const& value : cell_range) {
+      ARROW_RETURN_NOT_OK(ConversionTraits<ValueCType>::AppendRow(value_builder, value));
     }
     return Status::OK();
   }
+};
+
+template <typename ValueCType, typename BuilderType, typename Range>
+struct AppendMultipleRowsHelper<
+    ValueCType, BuilderType, Range,
+    typename std::enable_if<HasAppendMultipleRows<
+        ValueCType, ElementBuilderType<ValueCType>, Range>::value>::type> {
+  static Status impl(BuilderType& builder, Range&& cell_range) {
+    ARROW_RETURN_NOT_OK(builder.Append());
+    auto& value_builder =
+        ::arrow::internal::checked_cast<ElementBuilderType<ValueCType>&>(
+            *builder.value_builder());
+    return ConversionTraits<ValueCType>::AppendMultipleRows(value_builder, cell_range);
+  }
+};
+}  // namespace internal
+
+/// Add multiple rows to a (large) list builder.
```

Comment:
This is used to append a single value to a list builder; it appends multiple values to the child builder. Please update the comment and rename this function accordingly

Reply:
Will do. I named it to be consistent with AppendRow, as it also appends to child builder in current implementation. How about `AppendMultipleValues`?

Reply:
Since this is list specific, maybe AppendListValue?

Reply:
Renamed this helper function as `AppendCellRange` and free function provided for specializaton as `AppendListValues`. Does it sound right?